### PR TITLE
[Xharness] Do use the traits provided by the mono sdk.

### PR DESCRIPTION
**Pull request by mandel-macaque**
Rather than hardcoding the tests ignored categories and traits, use the
ones that are provided by the mono sdk for each runner.

Ps: created as a draft because I believe that there is a mismatch between the traits/categories present in mono and the ones we are using. We have some missing which will result in tests failing.

### CODE REVIEW

**Comment by akoeplinger** [comment]
Fixing the nunit excludes is in progress at https://github.com/mono/mono/pull/14695

File: tools/bcl-test-importer/BCLTestImporter/BCLTestProjectGenerator.cs [hunk 1]
```diff
@@ -445,6 +445,44 @@ internal static IEnumerable<string> GetIgnoreFiles (string templatePath, string
...
+				rootPath = Path.Combine (rootPath, "ios-bcl", "monotouch_watch", "tests");
+				break;
+			case Platform.MacOSFull:
+				rootPath = Path.Combine (rootPath, "mac-bcl", "xammac_net_4_5", "tests");
+				break;
+			case Platform.MacOSModern:
+				rootPath = Path.Combine (rootPath, "mac-bcl", "xammac", "tests");
+				break;
+			}
+			foreach (var fileName in new [] {"nunit-excludes.txt", "xunit-excludes.txt"}) {
```
Comment (rolfbjarne) [inline tools/bcl-test-importer/BCLTestImporter/BCLTestProjectGenerator.cs]: There's no need for an iterator, you can simply return an array:

```csharp
return new [] {
    Path.Combine (rootPath, "nunit-excludes.txt"),
    Path.Combine (rootPath, "xunit-excludes.txt")
};
```

which is faster and uses less memory.

File: tests/bcl-test/BCLTests/templates/common/IgnoreFileParser.cs [hunk 2]
```diff
@@ -70,6 +70,21 @@ public static async Task<IEnumerable<string>> ParseContentFilesAsync (string con
 			return ignoredTests;
 		}
 		
+		public static async Task<IEnumerable<string>> ParseTraitsContentFileAsync (string contentDir, bool isXUnit)
+		{
+			var ignoredTraits = new List<string> ();
+			var ignoreFile = Path.Combine (contentDir, isXUnit ? "xunit-excludes.txt" : "nunit-excludes.txt");
+			using (var reader = new StreamReader (ignoreFile)) {
+				string line;
+				while ((line = await reader.ReadLineAsync()) != null) {
```
Comment (rolfbjarne) [inline tests/bcl-test/BCLTests/templates/common/IgnoreFileParser.cs]:
```suggestion
				while ((line = await reader.ReadLineAsync ()) != null) {
```

File: tools/bcl-test-importer/BCLTestImporter/BCLTestProjectGenerator.cs (hunk 1 above)
Comment (mandel-macaque) [reply tools/bcl-test-importer/BCLTestImporter/BCLTestProjectGenerator.cs]: that was the python developer in me, I'll fix that :)

**Comment by mandel-macaque** [comment]
Looks like the blocking tests is due to some ignored categories missing on the mono side.

**Comment by mandel-macaque** [comment]
All tests are known:

* MSBbuild issues: https://github.com/mono/mono/issues/14876 and https://github.com/mono/mono/issues/14875
* Mono Mac OS X BCL tests group 4/Mac/Modern: Failed (Test run crashed (exit code: 134).) https://github.com/xamarin/maccore/issues/1685
* [NUnit] Mono CorlibTests/watchOS 32-bits - simulator/Debug: Crashed https://github.com/xamarin/maccore/issues/581
* [xUnit] Mono SystemCoreXunit/watchOS 32-bits - simulator/Debug: Crashed https://github.com/xamarin/maccore/issues/581

**Pull request merged**